File: src/types/events.ts
/**
 * Event & Recurrence Types
 */

export interface EventCategory {
  id: number;
  name: string;
  created_at: string;
  updated_at: string;
}

export type RecurrenceFrequency = 'daily' | 'weekly' | 'monthly' | 'yearly';
export type RecurrenceEndType = 'never' | 'count' | 'date';

export interface RecurrenceRule {
  id: number;
  frequency: RecurrenceFrequency;
  interval: number;
  day_of_week: number | null;
  day_of_month: number | null;
  end_type: RecurrenceEndType;
  end_count: number | null;
  end_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface RecurrenceRuleCreate {
  frequency: RecurrenceFrequency;
  interval?: number;
  day_of_week?: number | null;
  day_of_month?: number | null;
  end_type?: RecurrenceEndType;
  end_count?: number | null;
  end_date?: string | null;
}

export interface Event {
  id: number;
  name: string;
  date: string;
  start_time: string | null;
  end_time: string | null;
  location: string | null;
  description: string | null;
  category_id: number | null;
  recurrence_rule_id: number | null;
  created_at: string;
  updated_at: string;
  is_occurrence?: boolean;
  master_id?: number | null;
  occurrence_date?: string | null;
}

export interface EventCreate {
  name: string;
  date: string;
  start_time?: string | null;
  end_time?: string | null;
  location?: string | null;
  description?: string | null;
  category_id?: number | null;
  recurrence_rule_id?: number | null;
}

export interface EventUpdate extends Partial<EventCreate> {}

File: backend/app/schemas/events.py
"""
Pydantic schemas for Events API.
"""

import datetime
import re
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field, field_validator

# Time format regex: HH:MM (24-hour format)
TIME_PATTERN = re.compile(r'^([01]\d|2[0-3]):([0-5]\d)$')


class EventBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    date: datetime.date
    start_time: Optional[str] = Field(None, max_length=10)
    end_time: Optional[str] = Field(None, max_length=10)
    location: Optional[str] = Field(None, max_length=500)
    description: Optional[str] = Field(None, max_length=5000)
    category_id: Optional[int] = None
    recurrence_rule_id: Optional[int] = None

    @field_validator('start_time', 'end_time')
    @classmethod
    def validate_time_format(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if not TIME_PATTERN.match(v):
            raise ValueError('Time must be in HH:MM format (24-hour)')
        return v


class EventCreate(EventBase):
    pass


class EventUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    date: Optional[datetime.date] = None
    start_time: Optional[str] = Field(None, max_length=10)
    end_time: Optional[str] = Field(None, max_length=10)
    location: Optional[str] = Field(None, max_length=500)
    description: Optional[str] = Field(None, max_length=5000)
    category_id: Optional[int] = None
    recurrence_rule_id: Optional[int] = None

    @field_validator('start_time', 'end_time')
    @classmethod
    def validate_time_format(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if not TIME_PATTERN.match(v):
            raise ValueError('Time must be in HH:MM format (24-hour)')
        return v


class EventResponse(EventBase):
    id: int
    created_at: datetime.datetime
    updated_at: datetime.datetime

    model_config = ConfigDict(from_attributes=True)


class EventOccurrenceResponse(EventResponse):
    """Event response with occurrence metadata for recurring events."""
    is_occurrence: bool = False  # True if this is a virtual occurrence, not the master
    master_id: Optional[int] = None  # ID of the master event (if is_occurrence=True)
    occurrence_date: Optional[datetime.date] = None  # The date of this specific occurrence
